daemon/firewall/firewall_manager.go

```diff
@@ -0,0 +1,502 @@
+	// of line number calculation(we ignore everything but the 'nordvpn-<priority>' comment or the lack of thereof).
+	args := "-L " + chain.String() + " --numeric"
+
+	output, err := i.cmdRunner.runCommand(command, args)
+	if err != nil {
+		return 0, fmt.Errorf("listing iptables rules: %w", err)
+	}
+
+	// Skip first two lines of output they are the chain name and table values name.
+	outputLines := strings.Split(string(output), "\n")[2:]
+
+	if len(outputLines) == 0 {
+		return 1, nil
+	}
+
+	nordvpnCommentPattern := regexp.MustCompile(`nordvpn-(\d+)`)
+
+	lastNordvpnRuleLine := 0
+	for ruleIndex, line := range outputLines {
+		matches := nordvpnCommentPattern.FindStringSubmatch(line)
+		if len(matches) > 0 {
+			// Offset by 1 because iptables rules are 1 based
+			lastNordvpnRuleLine = ruleIndex + 1
+		}
+	}
+
+	// Rules are present in iptables, but there are no nordvpn rules. In this case we want to insert the rule at the top
+	// so that it will be prioritized over other rules.
+	if len(outputLines) > 0 && lastNordvpnRuleLine == 0 {
+		return 1, nil
+	}
+
+	// get next lowest index
+	for lineIndex, line := range outputLines[:lastNordvpnRuleLine] {
+		if strings.Contains(line, priority.toComment()) {
+			return lineIndex + 1, nil
+		}
+
+		matches := nordvpnCommentPattern.FindStringSubmatch(line)
+		if len(matches) < 2 {
+			continue
+		}
+
+		rulePriority, err := strconv.Atoi(matches[1])
+		if err != nil {
+			return 0, fmt.Errorf("converting priority to int: %w", err)
+		}
+
+		if rulePriority < int(priority) {
+			// Array indexes are 0 based, iptables line numbers are 1 based.
+			return lineIndex + 1, nil
+		}
+	}
+
+	// Offset by 1 because we want to insert the rule after the last discovered nordvpn rule.
+	return lastNordvpnRuleLine + 1, nil
+}
+
+func (i iptablesManager) insertRule(rule FwRule) error {
+	return i.executeCommand(true, rule)
+}
+
+func (i iptablesManager) deleteRule(rule FwRule) error {
+	return i.executeCommand(false, rule)
+}
+
+type PortRange struct {
+	min int
+	max int
+}
+
+type FwRule struct {
+	chain    iptablesChain
+	version  ipVersion
+	params   string
+	priority rulePriority
+}
+
+func NewFwRule(chain iptablesChain, version ipVersion, params string, priority rulePriority) FwRule {
+	return FwRule{
+		chain:    chain,
+		version:  version,
+		params:   params,
+		priority: priority,
+	}
+}
+
+// ToInsertAppendCommand returns the FwRule converted to insert command(-I <CHAIN> <ARGS>) or append command if index is
+// -1.
+func (f FwRule) ToInsertAppendCommand(index int) string {
+	return fmt.Sprintf("-I %s %d %s %s", f.chain, index, f.params, f.priority.toCommentArgs())
+}
+
+func (f FwRule) ToDeleteCommand() string {
+	return fmt.Sprintf("-D %s %s %s", f.chain, f.params, f.priority.toCommentArgs())
+}
+
+type FirewallManager struct {
+	iptablesManager   iptablesManager
+	devices           device.ListFunc // list network interfaces
+	allowlistRules    []FwRule
```

Comment:
We need to keep track of the actual rules for the `enable`/`disable` functionality. When firewall is disabled we still want to store the rules so that they can be re-applied once it is enabled.